File: backend/app/routes/animations.py
from fastapi import APIRouter
from pydantic import BaseModel

from app.models.animation import AnimationClip, AnimationJob
from app.services import animation_service

router = APIRouter(prefix="/api/animations", tags=["animations"])


class CreateAnimationJobRequest(BaseModel):
    source_rig_job_id: str
    clip_id: str


@router.get("/clips", response_model=list[AnimationClip])
async def list_clips():
    return animation_service.list_clips()


@router.post("/create", response_model=AnimationJob, status_code=201)
async def create_animation_job(body: CreateAnimationJobRequest):
    return animation_service.create_animation_job(body.source_rig_job_id, body.clip_id)


@router.get("", response_model=list[AnimationJob])
async def list_animation_jobs():
    return animation_service.list_animation_jobs()


@router.get("/{animation_job_id}", response_model=AnimationJob)
async def get_animation_job(animation_job_id: str):
    return animation_service.get_animation_job(animation_job_id)

File: frontend/src/lib/animations.ts
export interface AnimationClip {
  id: string;
  name: string;
  clip_type: string;
  duration_seconds: number;
  compatible_rig_types: string[];
  description: string;
}

export interface AnimationJob {
  id: string;
  source_rig_job_id: string;
  status: "queued" | "processing" | "completed" | "failed";
  clip_id: string;
  provider: string;
  output_files: string[];
  message: string;
  error: string | null;
  created_at: string;
  updated_at: string;
}

export async function getAnimationClips(): Promise<AnimationClip[]> {
  const res = await fetch("/api/animations/clips");
  if (!res.ok) throw new Error("Failed to fetch animation clips");
  return res.json();
}

export async function createAnimationJob(
  source_rig_job_id: string,
  clip_id: string
): Promise<AnimationJob> {
  const res = await fetch("/api/animations/create", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ source_rig_job_id, clip_id }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(
      (err as { detail?: string }).detail ?? "Failed to create animation job"
    );
  }
  return res.json();
}

export async function getAnimationJob(
  animation_job_id: string
): Promise<AnimationJob> {
  const res = await fetch(`/api/animations/${animation_job_id}`);
  if (!res.ok) throw new Error("Failed to fetch animation job");
  return res.json();
}

export async function listAnimationJobs(): Promise<AnimationJob[]> {
  const res = await fetch("/api/animations");
  if (!res.ok) throw new Error("Failed to fetch animation jobs");
  return res.json();
}
